### PR TITLE
[GH-55] Add realtime metric support.

### PR DESCRIPTION
**Pull request by jealous**
Add subscription support for realtime metric.

Details:
* List metric resources.
* List/Create/Delete metric real time query.
* Query metric query result.
* Add `RepeatedTimer` utility for management of repeated tasks.
* List disks.
* Add calculators meta info manager.
* Enable/Disable real time metric collection.
* Add metric properties in printed string of resource/resource list.
* Add basic persist stats to csv support.

Add following metrics:
* disk
  * read IOPS
  * write IOPS
  * read bandwidth
  * write bandwidth
  * utilization
* lun
  * read IOPS
  * write IOPS
  * read bandwidth
  * write bandwidth
  * utilization
* filesystem
  * read IOPS
  * write IOPS
  * read bandwidth
  * write bandwidth
* storage processor
  * net in bandwidth
  * net out bandwidth
  * block read IOPS
  * block write IOPS
  * block read bandwidth
  * block write bandwidth
  * CIFS read IOPS
  * CIFS write IOPS
  * CIFS read bandwidth
  * CIFS write bandwidth
  * NFS read IOPS
  * NFS write IOPS
  * NFS read bandwidth
  * NFS write bandwidth
  * utilization
  * block cache read hit ratio
  * block cache write hit ratio

### CODE REVIEW

File: storops/lib/common.py [hunk 1]
```diff
@@ -413,18 +413,29 @@ def is_valid(value):
...
-    lock_folder = path.join(get_local_folder(), 'file_lock')
-    if not path.exists(lock_folder):
+def assure_folder(folder):
+    if not path.exists(folder):
         try:
-            makedirs(lock_folder)
+            makedirs(folder)
         except OSError as e:
-            # ignore existed error
+            # ignore exited error
```
Comment (Tianqi-Tang) [inline storops/lib/common.py]: existing

File: storops/unity/calculator.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+    def _div(cls, op1, op2):
+        if op1 is None or op2 is None or op2 == 0:
+            r = cls.NaN
+        else:
+            r = op1 / op2
+        return r
+
+    def __div__(self, other):
+        nan = self.NaN
+        if other == 0:
```
Comment (Tianqi-Tang) [inline storops/unity/calculator.py]: Line 105-107 is conflict with login in 97-98

File: storops/unity/calculator.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+            raise ZeroDivisionError()
+        elif other is None:
+            raise ValueError('divide by none is not allowed.')
+        else:
+            ret = self._apply_op(nan, self._div, other)
+        return ret
+
+    def __rtruediv__(self, other):
+        if other is None:
+            raise ValueError('cannot use none as dividend.')
```
Comment (Tianqi-Tang) [inline storops/unity/calculator.py]: ditto

File: README.rst [hunk 4]
```diff
@@ -94,10 +94,52 @@ Feature List
         - list/create/delete NFS shares
         - list/create/delete DNS servers
         - list ip ports.
+        - list/create/delete link aggregation
```
Comment (jealous) [inline README.rst]: Done.

File: README.rst [hunk 5]
```diff
@@ -94,10 +94,52 @@ Feature List
         - list/create/delete NFS shares
         - list/create/delete DNS servers
         - list ip ports.
+        - list/create/delete link aggregation
+        - list/create/delete Consistency Groups
+        - list/create/delete metric real time query
+        - list metric query result
```
Comment (jealous) [inline README.rst]: Done.

File: storops/lib/common.py (hunk 1 above)
Comment (jealous) [inline storops/lib/common.py]: Done.

File: storops/unity/calculator.py (hunk 2 above)
Comment (jealous) [inline storops/unity/calculator.py]: Done.

File: storops/unity/calculator.py (hunk 3 above)
Comment (jealous) [inline storops/unity/calculator.py]: Done.

File: storops/unity/calculator.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
+# coding=utf-8
+# Copyright (c) 2015 EMC Corporation.
```
Comment (jealous) [inline storops/unity/calculator.py]: Done.

File: storops/unity/resource/disk.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2015 EMC Corporation.
```
Comment (jealous) [inline storops/unity/resource/disk.py]: Done.

File: storops/unity/resource/metric.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
+# coding=utf-8
+# Copyright (c) 2015 EMC Corporation.
```
Comment (jealous) [inline storops/unity/resource/metric.py]: Done.

File: storops/unity/resource/metric.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    def update(self, data=None):
+        super(UnityMetricQueryResultList, self).update(data)
+        self._path_result_map = {}
+
+    def by_path(self, path):
+        if not self._path_result_map:
+            for r in self:
+                self._path_result_map[r.path] = r
+        return self._path_result_map.get(path)
+        # filtered = list(filter(lambda r: r.path == path, self))
```
Comment (jealous) [inline storops/unity/resource/metric.py]: Yes, removed.  Thanks.

**Pull request merged**